semgrep/semgrep/evaluation.py

```diff
@@ -391,12 +409,30 @@ def evaluate(
     return output, steps_for_debugging
 
 
-def interpolate_message_metavariables(rule: Rule, pattern_match: PatternMatch) -> str:
+def interpolate_message_metavariables(
+    rule: Rule,
+    pattern_match: PatternMatch,
+    all_pattern_match_metavariables: Dict[str, List[PatternMatch]],
+) -> str:
     msg_text = rule.message
-    for metavar in pattern_match.metavars:
-        msg_text = msg_text.replace(
-            metavar, pattern_match.get_metavariable_value(metavar)
-        )
+    for metavar_text in all_pattern_match_metavariables:
+        replace_text = metavar_text
+        try:  # Always prefer the pattern match metavariable first.
+            replace_text = pattern_match.get_metavariable_value(metavar_text)
+        except KeyError:  # If one isn't present, retrieve the value from all metavariables.
+            pattern_matches_with_metavars_that_enclose_match: List[PatternMatch] = list(
```

Comment:
This seems ok but feels very fragile. Is there a way we can have a pattern_match object contain a list of all metavars contained within it then in interpolation we can just do a lookup?

Reply:
Potentially: we might be able to populate sibling pattern nodes with all matched metavariables.

But then, I thought that wouldn't be so much more efficient than gathering up all the metavariables for all the patterns. 🤷‍♂️